File: unit4/DSA/backtracking/wordInMatrix.js
function exist(board, word) {
    const n = board.length;
    const m = board[0].length;
    const directions = [
        [0, 1],
        [0, -1],
        [1, 0],
        [-1, 0]
    ];
    function dfs(x, y, idx) {
        if (idx === word.length) {
            return true;
        }
        if (x < 0 || x >= n || y < 0 || y >= m || board[x][y] !== word[idx]) {
            return false;
        }
        const temp = board[x][y];
        board[x][y] = '#';
        for (let [dx, dy] of directions) {
            if (dfs(x + dx, y + dy, idx + 1)) {
                return true;
            }
        }
        board[x][y] = temp;
        return false;
    }
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (dfs(i, j, 0)) {
                return true;
            }}}
    return false;
}
let n = 3, m = 4;
let board = [
    ['A', 'B', 'C', 'E'],
    ['S', 'F', 'C', 'S'],
    ['A', 'N', 'E', 'E']
];
let word = "SEEN";
console.log(exist(board, word)); // Output: true

File: unit4/DSA/backtracking/practce/wordInmatrix.py
def wordMatrix(n,m,board,word):

    def DFS(i,j,index):
        if index ==len(word):
            return True
        if i>=n or i<0 or j>=m or j<0 or board[i][j] !=word[index]:
            return False
        temp = board[i][j]
        board[i][j] = "#"
        for x,y in direction:
            if DFS(i+x , j+y,  index+1):
                return True

        board[i][j] = temp
        return False    


    direction = [
        [0,1],[0,-1],[1,0],[-1,0]   # east west  south nnoth
    ]
    for i in range(n):
        for j in range(m):
            if DFS(i,j,0):
                return True
    return False        


n,m = 3,4
board=[
    ['A', 'B', 'C', 'E'],
    ['S', 'F', 'C', 'S'],
    ['A', 'N', 'E', 'E']
]
word="SEEN"
print(wordMatrix(n,m,board,word))
